golem-cli/src/service/deploy.rs

```diff
@@ -0,0 +1,195 @@
+// Copyright 2024 Golem Cloud
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use crate::model::deploy::TryUpdateAllWorkersResult;
+use crate::model::{Format, GolemError, GolemResult, WorkerName, WorkerUpdateMode};
+use crate::service::component::ComponentService;
+use crate::service::worker::WorkerService;
+use async_trait::async_trait;
+use golem_common::model::{ComponentId, WorkerId};
+use golem_common::uri::oss::uri::ComponentUri;
+use golem_common::uri::oss::urn::WorkerUrn;
+use inquire::Confirm;
+use std::fmt::Display;
+use std::sync::Arc;
+use tracing::{debug, info};
+
+/// Higher-level deployment operations implemented on top of the underlying services
+#[async_trait]
+pub trait DeployService {
+    type ProjectContext: Send + Sync;
+
+    async fn try_update_all_workers(
+        &self,
+        component_uri: ComponentUri,
+        project: Option<Self::ProjectContext>,
+        mode: WorkerUpdateMode,
+    ) -> Result<GolemResult, GolemError>;
+
+    async fn redeploy(
+        &self,
+        component_uri: ComponentUri,
+        project: Option<Self::ProjectContext>,
+        non_interactive: bool,
+        format: Format,
+    ) -> Result<GolemResult, GolemError>;
+}
+
+pub struct DeployServiceLive<ProjectContext> {
+    pub component_service: Arc<dyn ComponentService<ProjectContext = ProjectContext> + Send + Sync>,
+    pub worker_service: Arc<dyn WorkerService<ProjectContext = ProjectContext> + Send + Sync>,
+}
+
+#[async_trait]
+impl<ProjectContext: Display + Send + Sync> DeployService for DeployServiceLive<ProjectContext> {
+    type ProjectContext = ProjectContext;
+
+    async fn try_update_all_workers(
+        &self,
+        component_uri: ComponentUri,
+        project: Option<Self::ProjectContext>,
+        mode: WorkerUpdateMode,
+    ) -> Result<GolemResult, GolemError> {
```

Comment:
it could be better if we had api at worker service, which will process it at server side, but probably ok for now

Reply:
I don't agree with that, these are higher level workflows built on the worker service API, including the need to enumerate through all workers through the "slow" worker enumeration. I would not put it in the service side.